### PR TITLE
checkpoint_abort not working

### PR DESCRIPTION
**Pull request by prakashcv13**
#### Bug/feature Description
* checkpoint_abort is not working.
* Steps to reproduce:

           1. Set preempt_order to "C" in sched_config.
           2. Setup only one node in the complex having ncpus=4.
           3. Setup action for checkpoint_abort and restart in mom config.
           4. Have one normal priority and one express queue.
           5. Submit a job to the normal priority queue, requesting 3 ncpus confirm it starts running.
           6. Submit a job to the express queue, requesting 4 ncpus.

Expected Behavior: Job submitted to express queue should preempt the job submitted to the normal queue and starts running. The job submitted to the normal queue goes to the queued state 'Q'.
Observed Behavior: Job submitted to the normal queue remains in 'E' state for a while and then moved to hold state 'H'. Job submitted to the express queue never runs.
Logs - [steps to reproduce.txt](https://github.com/PBSPro/pbspro/files/2138508/steps.to.reproduce.txt)

#### Affected Platform(s)
* All

#### Cause / Analysis / Design
* This [fix](https://github.com/PBSPro/pbspro/pull/412) moved the setting of job substate to rerun from req_jobhold() to post_hold(). 
* Due to this, when MoM sends an obit for the preempted job, server was not sending the acknowledgement to the MoM as the logic falls under "rerun" block in job_obit().

#### Solution Description
* Moved the logic to send acknowledgement to MoM to the non-rerun block

#### Testing logs/output
* [testing_logs_after_fix.txt](https://github.com/PBSPro/pbspro/files/2138517/testing_logs_after_fix.txt)

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

**Comment by prakashcv13** [comment]
@arungrover - yes that is what is happening - and after that -
- scheduler releases the hold
- server sets the substate to queued (10).

**Comment by lisa-altair** [comment]
@prakashcv13 please add a PTL test for this bug fix.

File: src/server/req_jobobit.c [hunk 1]
```diff
@@ -2269,24 +2288,6 @@ job_obit(struct resc_used_update *pruu, int stream)
 			(JOB_SVFLG_CHKPT | JOB_SVFLG_ChkptMig)) == 0) {
 			job_attr_def[(int)JOB_ATR_resc_used].at_free(
 				&pjob->ji_wattr[(int)JOB_ATR_resc_used]);
-
-		} else if ((pjob->ji_qs.ji_svrflags & JOB_SVFLG_CHKPT) &&
```
Comment (arungrover) [inline src/server/req_jobobit.c]: Are you sure we don't need this part of the code here? What if checkpoint_abort fails on mom and in post_hold we do not remove CHKPT flag from the job and then the job exits. I think in that case control will come here and will not reply ack to mom.
Since this is a timing issue, have you thought of order of operation where server processes obit before processing holdjob response (when job is ending at the same time it is getting held). These are some tricky scenarios and you may want to try some of those.

**Comment by lisa-altair** [comment]
Also, since the culprit you point to (PP-1017) was for a Cray.  Can you please also test checkpointing with compute nodes on a Cray?  I'm worried that the ALPS reservation isn't going to be released.  So please run the second job (that causes the checkpoint) on the same compute node.

File: src/server/req_jobobit.c (hunk 1 above)
Comment (prakashcv13) [reply src/server/req_jobobit.c]: I had tested with the checkpoint_abort script returning 1 instead of 0. Checkpointing fails, but MoM decides to send an Obit that the server responds and the job is preempted.
line 2269 is when substate is rerun and checkpoint flag is NOT set.
I do not think this is a timing issue, PP-1017 was.
I believe, the testing I have done is enough, if there is a specific scenario you have in your mind, please let me know, will be happy to test it.

**Comment by prakashcv13** [comment]
@lisa-altair and @arungrover - thank you for the review and the comments. I have added a PTL test case to this PR. The testing was done on Cray and local Linux. logs for the local Linux testing are attached.

Thanks again!
[ptl_local_linux.txt](https://github.com/PBSPro/pbspro/files/2143039/ptl_local_linux.txt)

File: test/tests/functional/pbs_checkpoint.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+from tests.functional import *
+
+
+class TestCheckpoint(TestFunctional):
+    """
+    This test suite targets Checkpoint functionality.
+    """
+
+    def num_compute_vnodes(self):
```
Comment (lisa-altair) [inline test/tests/functional/pbs_checkpoint.py]: This same function also exists in pbs_cray_suspend_resume.py, please put this in the framework  fw/ptl/utils/pbs_crayutils.py.

File: test/tests/functional/pbs_checkpoint.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        self.du.chmod(path=abort_file, mode=0755)
+        c = {'$action': 'checkpoint_abort 30 !' + abort_file + ' %sid'}
+        self.mom.add_config(c)
+        self.scheduler.set_sched_config({'preempt_order': 'C'})
+        a = {'queue_type': 'e',
+             'started': 'True',
+             'enabled': 'True',
+             'Priority': 200}
+        self.server.manager(MGR_CMD_CREATE, QUEUE, a, "expressq")
+        self.platform = DshUtils().get_platform()
```
Comment (lisa-altair) [inline test/tests/functional/pbs_checkpoint.py]: I think this should be self.du.get_platform()

File: test/tests/functional/pbs_checkpoint.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            node_attrs = self.server.status(NODE, id=self.server.shortname)[0]
+            ncpus = node_attrs['resources_available.ncpus']
+            attrs = {'Resource_List.select': '1:ncpus=' + ncpus}
+        else:
+            nv = self.num_compute_vnodes()
+            attrs = {ATTR_l + '.select': '%d:ncpus=1' % nv,
+                     ATTR_l + '.place': 'scatter'}
+
+        j1 = Job(TEST_USER, attrs)
+        j1.set_sleep_time(5)
```
Comment (lisa-altair) [inline test/tests/functional/pbs_checkpoint.py]: Depending on the speed of the system, isn't it possible j1 could be finished before j2 can cause it to be checkpointed.
Also, if the test is running on a Cray system, will this call affect the Cray default job's aprun sleep of 100 sec?

File: test/tests/functional/pbs_checkpoint.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        jid1 = self.server.submit(j1)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid1)
+        attrs['queue'] = 'expressq'
+        j2 = Job(TEST_USER, attrs)
+        j2.set_sleep_time(1)
+        jid2 = self.server.submit(j2)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid2)
+        self.server.expect(JOB, {'job_state': 'Q'}, id=jid1)
+        self.server.delete(id=jid1)
```
Comment (lisa-altair) [inline test/tests/functional/pbs_checkpoint.py]: Since you're giving the jobs such short sleep times, can't you let both jobs finish normally?  That way we should see j2 end and j1 run again and then also end without errors.

File: src/server/req_jobobit.c (hunk 1 above)
Comment (lisa-altair) [reply src/server/req_jobobit.c]: > I had tested with the checkpoint_abort script returning 1 instead of 0. Checkpointing fails, but MoM decides to send an Obit that the server responds and the job is preempted.

I'm not sure I understand what you're saying.  If preempt_order is "C", and the Checkpoint fails, how does the server respond that the job is preempted...didn't preemption fail?

File: src/server/req_jobobit.c [hunk 6]
```diff
@@ -2251,6 +2251,25 @@ job_obit(struct resc_used_update *pruu, int stream)
...
+			ack_obit(stream, pjob->ji_qs.ji_jobid);
+			pjob->ji_qs.ji_svrflags |= JOB_SVFLG_HASRUN;
+			svr_evaljobstate(pjob, &newstate, &newsubst, 1);
+			(void)svr_setjobstate(pjob, newstate, newsubst);
+			if (pjob->ji_mom_prot == PROT_TCP)
+				svr_disconnect(pjob->ji_momhandle);
+
+			pjob->ji_momhandle = -1;
+			pjob->ji_mom_prot = PROT_INVALID;
+			return;
```
Comment (lisa-altair) [inline src/server/req_jobobit.c]: Why do we return here?  Wouldn't we want to continue in the code and set a work task?

File: test/tests/functional/pbs_checkpoint.py (hunk 2 above)
Comment (prakashcv13) [reply test/tests/functional/pbs_checkpoint.py]: OK.

File: test/tests/functional/pbs_checkpoint.py (hunk 3 above)
Comment (prakashcv13) [reply test/tests/functional/pbs_checkpoint.py]: right.

File: test/tests/functional/pbs_checkpoint.py (hunk 4 above)
Comment (prakashcv13) [reply test/tests/functional/pbs_checkpoint.py]: makes sense.

File: test/tests/functional/pbs_checkpoint.py (hunk 5 above)
Comment (prakashcv13) [reply test/tests/functional/pbs_checkpoint.py]: that would be testing if the restart works well and according to me that is out of scope of this bug.

File: src/server/req_jobobit.c (hunk 1 above)
Comment (prakashcv13) [reply src/server/req_jobobit.c]: the MoM sends an Obit after the checkpointing fails, the server responds to it, and the job gets preempted. The pre-emption goes well regardless of whether checkpointing was successful because the MoM is sending the Obit.

File: src/server/req_jobobit.c (hunk 6 above)
Comment (prakashcv13) [reply src/server/req_jobobit.c]: No we don't need it. The job needs to be requeued.

File: src/server/req_jobobit.c (hunk 1 above)
Comment (lisa-altair) [reply src/server/req_jobobit.c]: @prakashcv13 after digging through the code with Bhroam, I believe you are only seeing this behavior because you killed all processes but returned a non-zero exit.  Since there nothing else for the job to do it ends, and that's why you see the Obit.  What happens if your checkpoint script only returns non-zero exit (need a long enough job that it also doesn't end by then)?  According to what we saw in the code, Mom should be sending a req_reject and not an Obit.

Comment (bhroam) [reply src/server/req_jobobit.c]: Just checking for the job in the Q state doesn't tell if the job was checkpointed.  You need to see if the exec_vnode attribute is also set.  In any case, this seems wrong.  If the checkpoint fails, then the qhold should be rejected.  Could you test this by hand?  Set up checkpoint where the checkpoint_abort script just returns 1.  Submit a job and then qhold it.  If it goes into the H state, something strange is going on.

Comment (prakashcv13) [reply src/server/req_jobobit.c]: @lisa-altair and @bhroam - I tried this scenario and you guys are right that checkpointing will fail and so will preemption. But, isn't that an expected behavior?
What we are trying to fix here is what went wrong after the fix of PP-1017. It was the missing ack to the obit which is added now. Rest of the things are as they were.
I believe what we have tested so far is enough.

File: test/tests/functional/pbs_checkpoint.py (hunk 4 above)
Comment (prakashcv13) [reply test/tests/functional/pbs_checkpoint.py]: I will increase the sleep time to 100 seconds.

**Comment by prakashcv13** [comment]
@lisa-altair - I have addressed your comments.

File: test/tests/functional/pbs_checkpoint.py (hunk 4 above)
Comment (bhroam) [reply test/tests/functional/pbs_checkpoint.py]: I believe the default sleep time is 100s.  If you have no reason to change it, I'd leave it as the default(don't call set_sleep_time()).  The main reason to change the sleep time is you want to wait until the jobs end.  I don't think that is the case here.

Comment (prakashcv13) [reply test/tests/functional/pbs_checkpoint.py]: To be on the safer side, I have increased it to 1000s.

**Comment by prakashcv13** [comment]
@arungrover, @bhroam, @lisa-altair - please find attached the testing logs.
[ptl_local_linux_0628.txt](https://github.com/PBSPro/pbspro/files/2146391/ptl_local_linux_0628.txt)

File: test/tests/functional/pbs_checkpoint.py (hunk 5 above)
Comment (prakashcv13) [reply test/tests/functional/pbs_checkpoint.py]: when I checked, the job was not deleted.

File: src/server/req_jobobit.c [hunk 7]
```diff
@@ -2251,6 +2251,25 @@ job_obit(struct resc_used_update *pruu, int stream)
 
 		check_block(pjob, "");		/* if block set, send word */
 
+		if ((pjob->ji_qs.ji_svrflags & JOB_SVFLG_CHKPT) &&
+			((pjob->ji_qs.ji_svrflags & JOB_SVFLG_SubJob) == 0)) {
+
+			/* non-migratable checkpoint, leave there */
+			/* and just requeue the job		      */
```
[the remark this answers is not in the document]
Comment (prakashcv13) [reply src/server/req_jobobit.c]: that is existing code which was moved, I will make it look standard.

File: src/server/req_jobobit.c (hunk 1 above)
Comment (prakashcv13) [reply src/server/req_jobobit.c]: If I do not kill the process, the pre-emption still happens. If the process is left lying around, it doesn't.

File: test/tests/functional/pbs_checkpoint.py (hunk 5 above)
Comment (bhroam) [reply test/tests/functional/pbs_checkpoint.py]: That is because you only have 1 test.  Jobs get deleted in setUp() of the following test.

File: src/server/req_jobobit.c (hunk 1 above)
Comment (bhroam) [reply src/server/req_jobobit.c]: I think this is a side effect of our communication model between mom and the server.  I think what is happening is the server says, "checkpoint this job".  The checkpoint abort script fails but kills the process.  There are no processes left for the job, so the mom notices this and sends the normal end of job obit to the server.  The server was expecting an obit from the checkpoint so it assumes this is where the obit is from.  Now we have a case where the checkpoint_abort script has failed, but we checkpoint the job anyway.  In any case, this isn't part of your bug fix.  Can you file it?  Thanks again for testing it.

File: test/tests/functional/pbs_checkpoint.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        j1.set_sleep_time(1000)
+        jid1 = self.server.submit(j1)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid1)
+        attrs['queue'] = 'expressq'
+        j2 = Job(TEST_USER, attrs)
+        j2.set_sleep_time(1)
+        jid2 = self.server.submit(j2)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid2)
+        self.server.expect(JOB, {'job_state': 'Q'}, id=jid1)
+        self.server.delete(id=jid1)
```
[the remark this answers is not in the document]
Comment (lisa-altair) [reply test/tests/functional/pbs_checkpoint.py]: @vccardenas I don't agree with this statement.  @prakashcv13 has already written a test that reproduces the bug.  I see no reason for him to write a test for every way that the bug can be reproduced...

[the remark this answers is not in the document]
Comment (iestockdale) [reply test/tests/functional/pbs_checkpoint.py]: The fix should be addressing all observed failure modes.  In order to show that, there need to be PTL tests for each of those.

[the remark this answers is not in the document]
Comment (bhroam) [reply test/tests/functional/pbs_checkpoint.py]: I agree with @lisa-altair.  There are three methods of triggering checkpointing.  All three send a checkpoint command to the mom.  Everything after that is the same.  @prakashcv13 changed code after the mom responds to the checkpoint command she was given.  He wrote a test which used one of the three methods of triggering checkpointing, since his fix has nothing to do with how one triggers checkpointing.  It has to do with how the job is checkpointed.

[the remark this answers is not in the document]
Comment (iestockdale) [reply test/tests/functional/pbs_checkpoint.py]: The reasons to have the PTL tests for the bug fix include ensuring that the assumptions behind the fix are correct and that all affected scenarios beneft from the fix.  The fix may be independent of the triggering method, but the consequences of the fix could be different (or not).  The way to show which it is with a test.

[the remark this answers is not in the document]
Comment (subhasisb) [reply test/tests/functional/pbs_checkpoint.py]: @iestockdale while i agree with you that we must have tests for the 3 different methods of checkpointing, i agree with the others that for this particular fix we do not need the contributor to cover those tests. What he has used to test well covers his area of change. I feel we should raise a separate bug to add test coverage in the other triggering methods, but that is beyond the scope of this fix alone.

File: src/server/req_jobobit.c (hunk 6 above)
Comment (lisa-altair) [reply src/server/req_jobobit.c]: Okay, thanks for explaining.

**Comment by lisa-altair** [comment]
@prakashcv13 thank you for moving num_compute_vnodes code to CrayUtils.  Earlier I mentioned that the same function also exists in pbs_cray_suspend_resume.py.  To finish creating a generic function, please update pbs_cray_suspend_resume.py to use the CrayUtils function instead of the local copy of the same function.  Thank you.

File: test/tests/functional/pbs_checkpoint.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+from tests.functional import *
+from ptl.utils.pbs_crayutils import CrayUtils
+
+
+class TestCheckpoint(TestFunctional):
+    """
+    This test suite targets Checkpoint functionality.
+    """
+
+    def test_checkpoint_abort_with_preemept(self):
```
Comment (vccardenas) [inline test/tests/functional/pbs_checkpoint.py]: @prakashcv13 , please add a docstring for this test.

File: test/fw/ptl/utils/pbs_crayutils.py [hunk 10]
```diff
@@ -175,3 +175,15 @@ def get_numthreads(self, nid):
...
+    def num_compute_vnodes(self):
+        """
+        Count the Cray compute nodes and return the value.
+        """
+        vlist = []
+        vnl = self.server.filter(
+            VNODE, {'resources_available.vntype': 'cray_compute'})
+        vlist = vnl["resources_available.vntype=cray_compute"]
+        nv = len(vlist)
+        self.assertNotEqual(nv, 0, "There are no cray_compute vnodes present.")
```
[the remark this answers is not in the document]
Comment (prakashcv13) [reply test/fw/ptl/utils/pbs_crayutils.py]: While I can change the docstring and the function name, I think having the utility function just to count the number of vnodes is sufficient even though it would have only one single return statement because there could be tests to confirm that the number of vnodes is equal to 0. So, I am leaving the decision to assert or not up to the user of the utility function.

File: test/tests/functional/pbs_checkpoint.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+        abort_script = """#!/bin/bash
+kill $1
+exit 0
+"""
+        abort_file = self.du.create_temp_file(body=abort_script)
+        self.du.chmod(path=abort_file, mode=0755)
```
[the remark this answers is not in the document]
Comment (prakashcv13) [reply test/tests/functional/pbs_checkpoint.py]: It doesn't seem like an issue to the MoM.

File: test/tests/functional/pbs_checkpoint.py [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+        c = {'$action': 'checkpoint_abort 30 !' + abort_file + ' %sid'}
+        self.mom.add_config(c)
+        self.scheduler.set_sched_config({'preempt_order': 'C'})
+        a = {'queue_type': 'e',
+             'started': 'True',
+             'enabled': 'True',
+             'Priority': 200}
+        self.server.manager(MGR_CMD_CREATE, QUEUE, a, "expressq")
```
[the remark this answers is not in the document]
Comment (prakashcv13) [reply test/tests/functional/pbs_checkpoint.py]: done!

File: test/tests/functional/pbs_checkpoint.py [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+        self.platform = self.du.get_platform()
+        if self.platform != 'cray' and self.platform != 'craysim':
+            node_attrs = self.server.status(NODE, id=self.server.shortname)[0]
+            ncpus = node_attrs['resources_available.ncpus']
+            attrs = {'Resource_List.select': '1:ncpus=' + ncpus}
```
[the remark this answers is not in the document]
Comment (prakashcv13) [reply test/tests/functional/pbs_checkpoint.py]: done!

File: test/tests/functional/pbs_checkpoint.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+        j1 = Job(TEST_USER, attrs)
+        j1.set_sleep_time(1000)
+        jid1 = self.server.submit(j1)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid1)
+        attrs['queue'] = 'expressq'
+        j2 = Job(TEST_USER, attrs)
+        j2.set_sleep_time(1)
+        jid2 = self.server.submit(j2)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid2)
+        self.server.expect(JOB, {'job_state': 'Q'}, id=jid1)
```
[the remark this answers is not in the document]
Comment (prakashcv13) [reply test/tests/functional/pbs_checkpoint.py]: removed the sleep time.

File: test/tests/functional/pbs_checkpoint.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (prakashcv13) [reply test/tests/functional/pbs_checkpoint.py]: thank you everyone!

File: test/tests/functional/pbs_checkpoint.py (hunk 9 above)
Comment (prakashcv13) [reply test/tests/functional/pbs_checkpoint.py]: added.

File: test/tests/functional/pbs_checkpoint.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (bhroam) [reply test/tests/functional/pbs_checkpoint.py]: Are you running pbs_benchpress as root?  I may be mistaken about the location of the file, but I'm pretty sure she makes sure the script is root owned.  PTL scripts are supposed to be able to be run as a non-root user.

**Comment by prakashcv13** [comment]
Thank you all for the review comments, I have addressed them. Please have a look.

File: src/server/req_jobobit.c [hunk 15]
```diff
@@ -2254,8 +2254,8 @@ job_obit(struct resc_used_update *pruu, int stream)
 		if ((pjob->ji_qs.ji_svrflags & JOB_SVFLG_CHKPT) &&
 			((pjob->ji_qs.ji_svrflags & JOB_SVFLG_SubJob) == 0)) {
 
-			/* non-migratable checkpoint, leave there */
-			/* and just requeue the job		      */
+			/* non-migratable checkpoint, leave there
+			 * and just requeue the job. */
```
Comment (bhroam) [inline src/server/req_jobobit.c]: Put the closing */ on the line below

File: test/fw/ptl/utils/pbs_crayutils.py [hunk 16]
```diff
@@ -176,14 +177,11 @@ def get_numthreads(self, nid):
...
-        vlist = []
-        vnl = self.server.filter(
-            VNODE, {'resources_available.vntype': 'cray_compute'})
+        vnl = server.filter(MGR_OBJ_NODE,
+                            {'resources_available.vntype': 'cray_compute'})
         vlist = vnl["resources_available.vntype=cray_compute"]
-        nv = len(vlist)
-        self.assertNotEqual(nv, 0, "There are no cray_compute vnodes present.")
-        return nv
+        return len(vlist)
```
Comment (bhroam) [inline test/fw/ptl/utils/pbs_crayutils.py]: Can you just put vnl[] in the len directly and get rid of vlist and still have the line under 79 characters?  Actually if you wanted to shrink the function even further, you could just return len(server.filter())

File: test/tests/functional/pbs_checkpoint.py [hunk 17]
```diff
@@ -44,40 +44,67 @@ class TestCheckpoint(TestFunctional):
...
+    cu = CrayUtils()
 
-    def test_checkpoint_abort_with_preemept(self):
+    def test_checkpoint_abort_with_preempt(self):
+        """
+        This test verifies that checkpoint_abort works as expected by
+        1. setting preempt_order in sched_config to 'C'.
+        2. submitting a job to the default queue.
+        3. submitting a job to the express queue.
+        4. confirming that the first job was checkpointed.
```
Comment (bhroam) [inline test/tests/functional/pbs_checkpoint.py]: Don't put the test steps in the doc string.  Just describe the test.

File: test/tests/functional/pbs_checkpoint.py [hunk 18]
```diff
@@ -44,40 +44,67 @@ class TestCheckpoint(TestFunctional):
...
         attrs['queue'] = 'expressq'
         j2 = Job(TEST_USER, attrs)
-        j2.set_sleep_time(1)
         jid2 = self.server.submit(j2)
         self.server.expect(JOB, {'job_state': 'R'}, id=jid2)
+
         self.server.expect(JOB, {'job_state': 'Q'}, id=jid1)
-        self.server.delete(id=jid1)
+        ck_dir = os.path.join(self.server.pbs_conf['PBS_HOME'], 'checkpoint/' +
+                              jid1 + '.CK')
```
Comment (bhroam) [inline test/tests/functional/pbs_checkpoint.py]: Let os.path.join() combine your path, don't add a string with a '/' in it.

File: test/tests/functional/pbs_checkpoint.py [hunk 19]
```diff
@@ -44,40 +44,67 @@ class TestCheckpoint(TestFunctional):
...
+        self.assertEqual(os.path.isdir(ck_dir), True,
+                         msg="Checkpoint directory for job not found")
```
Comment (bhroam) [inline test/tests/functional/pbs_checkpoint.py]: Use self.assertTrue()

File: test/fw/ptl/utils/pbs_crayutils.py (hunk 16 above)
Comment (prakashcv13) [reply test/fw/ptl/utils/pbs_crayutils.py]: len(server.filter()) would return 1. I will get rid of vlist by passing vnl[] to len().

File: test/tests/functional/pbs_checkpoint.py (hunk 19 above)
Comment (prakashcv13) [reply test/tests/functional/pbs_checkpoint.py]: done!

File: test/tests/functional/pbs_checkpoint.py (hunk 18 above)
Comment (prakashcv13) [reply test/tests/functional/pbs_checkpoint.py]: done!

File: test/tests/functional/pbs_checkpoint.py (hunk 17 above)
Comment (prakashcv13) [reply test/tests/functional/pbs_checkpoint.py]: ok, done!

**Comment by prakashcv13** [comment]
@bhroam - I have addressed your review comments - please have a look.

File: .gitignore [hunk 20]
```diff
@@ -193,6 +193,7 @@ src/tools/printjob.bin
 src/tools/printjob_svr.bin
 src/tools/tracejob
 src/tools/wrap_tcl.sh
+src/tools/pbs_sleep
```
Comment (bhroam) [inline .gitignore]: Check if this is already there.  I added pbs_sleep to the .gitignore file a couple of weeks ago.

File: test/tests/functional/pbs_checkpoint.py [hunk 21]
```diff
@@ -49,11 +49,9 @@ class TestCheckpoint(TestFunctional):
...
     def test_checkpoint_abort_with_preempt(self):
         """
-        This test verifies that checkpoint_abort works as expected by
-        1. setting preempt_order in sched_config to 'C'.
-        2. submitting a job to the default queue.
-        3. submitting a job to the express queue.
-        4. confirming that the first job was checkpointed.
+        This test verifies that checkpoint_abort works as expected when
+        preempt_order in sched_config is set to 'C' and a job in express
+        queue preempts a job in the default queue.
```
Comment (bhroam) [inline test/tests/functional/pbs_checkpoint.py]: I'd rephrase this.  How about 'when a job is preempted via checkpoint'.  That is what you are really testing.  There are other ways to cause it to happen than what you are doing which would test the same thing.  The real thing you are testing is preemption via checkpointing.

**Comment by bhroam** [comment]
Travis failed due to pep8

File: test/tests/functional/pbs_checkpoint.py [hunk 22]
```diff
@@ -44,40 +44,67 @@ class TestCheckpoint(TestFunctional):
...
+
         a = {'queue_type': 'e',
              'started': 'True',
              'enabled': 'True',
              'Priority': 200}
         self.server.manager(MGR_CMD_CREATE, QUEUE, a, "expressq")
+
         self.platform = self.du.get_platform()
         if self.platform != 'cray' and self.platform != 'craysim':
             node_attrs = self.server.status(NODE, id=self.server.shortname)[0]
```
Comment (lisa-altair) [inline test/tests/functional/pbs_checkpoint.py]: since the you changed how you form the job request, node_attrs is now no longer used.  Please remove it.

File: test/tests/functional/pbs_cray_suspend_resume.py [hunk 23]
```diff
@@ -48,6 +49,7 @@ class TestSuspendResumeOnCray(TestFunctional):
     This test suite expects the platform to be 'cray' and assumes that
     suspend/resume feature is enabled on it.
     """
+    cu = CrayUtils()
```
Comment (lisa-altair) [inline test/tests/functional/pbs_cray_suspend_resume.py]: Thanks for making these changes!

**Comment by lisa-altair** [comment]
@prakashcv13 please attach test logs from running your PTL test on a Linux system.  Thanks.

**Comment by prakashcv13** [comment]
@lisa-altair - attaching the logs.
[test_local_linux_0711.txt](https://github.com/PBSPro/pbspro/files/2190673/test_local_linux_0711.txt)

**Comment by lisa-altair** [comment]
Thanks @prakashcv13!  LGTM.

**Comment by prakashcv13** [comment]
@bhroam, @vccardenas - could you please let me know if there are further comments.

**Comment by prakashcv13** [comment]
@lisa-altair, @bhroam and @vcc, thank you for the review and approvals/sign-off. I have squashed my commits and rebased my branch. Request @PBSPro/pbspro-maintainers - to take it forward.

**Comment by bhroam** [comment]
I merged in the PR.
